File: libraries/OclIteratorOptimised.cpp
template<class T>
bool OclIteratorOptimised<T>::hasNext()
{
    bool result = false;
    if (position >= 0 && position < len)
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
bool OclIteratorOptimised<T>::hasPrevious()
{
    bool result = false;
    if (position > 1 && position <= len)
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
bool OclIteratorOptimised<T>::isAfterLast()
{
    bool result = false;
    if (position > len)
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
bool OclIteratorOptimised<T>::isBeforeFirst()
{
    bool result = false;
    if (position < 1)
    {
        result = true;
    }
    else
    {
        result = false;
    }
    return result;
}

template<class T>
int OclIteratorOptimised<T>::nextIndex()
{
    int result = 0;
    result = position + 1;
    return result;
}

template<class T>
int OclIteratorOptimised<T>::previousIndex()
{
    int result = 0;
    result = position - 1;
    return result;
}


template<class T>
void OclIteratorOptimised<T>::moveForward()
{
    int pre_position0 = position;
    //  if (position + 1 > elements->size())) { return; } 
    this->setPosition(pre_position0 + 1);
}

template<class T>
void OclIteratorOptimised<T>::moveBackward()
{
    int pre_position1 = position;
    //  if (position <= 1)) { return; } 
    this->setPosition(pre_position1 - 1);
}

template<class T>
void OclIteratorOptimised<T>::moveTo(int i)
{ this->setPosition(i); }

template<class T>
void OclIteratorOptimised<T>::movePosition(int i)
{
    int pos = this->getPosition();
    this->setPosition(i + pos);
}

template<class T>
void OclIteratorOptimised<T>::moveToStart()
{
    this->setPosition(0);
}

template<class T>
void OclIteratorOptimised<T>::moveToEnd()
{
    this->setPosition(this->len + 1);
}

template<class T>
void OclIteratorOptimised<T>::moveToFirst()
{
    this->setPosition(1);
}

template<class T>
void OclIteratorOptimised<T>::moveToLast()
{
    this->setPosition(this->len);
}

template<class T>
OclIteratorOptimised<T>* OclIteratorOptimised<T>::newOclIterator_Sequence(vector<T>* sq)
{
    OclIteratorOptimised<T>* ot = new OclIteratorOptimised<T>();
    int sze = sq->size(); 
    ot->elements = new T[sze]; 
    for (int i = 0; i < sze; i++)
    {
        ot->elements[i] = sq->at(i);
    }
    ot->setPosition(0);
    ot->len = sze; 
    ot->markedPosition = 0;
    return ot;
}

template<class T>
T OclIteratorOptimised<T>::getCurrent()
{
    T result = NULL;
    if (position < 1 || position > len)
    {
        return result;
    }
    result = elements[position - 1];
    return result;
}

template<class T>
void OclIteratorOptimised<T>::set(T x)
{
    this->elements[position - 1] = x;
}

template<class T>
void OclIteratorOptimised<T>::remove()
{
   this->elements[position - 1] = NULL;
}

template<class T>
T OclIteratorOptimised<T>::next()
{
    this->moveForward();
    return this->getCurrent();
}

template<class T>
T OclIteratorOptimised<T>::previous()
{
    this->moveBackward();
    return this->getCurrent();
}

template<class T>
T OclIteratorOptimised<T>::at(int i)
{
    return elements[i - 1];
}

template<class T>
int OclIteratorOptimised<T>::length()
{
    return this->len;
}

template<class T>
void OclIteratorOptimised<T>::close()
{
    position = 0;
    markedPosition = 0;
    elements = NULL;
    len = 0; 
}

File: libraries/ocliteratoroptimised.py

import ocl
import math
import re
import copy

from oclfile import *
from enum import Enum

# This is an optimised read-only iterator using tuples

class OclIterator :

  def __init__(self):
    self.position = 0
    self.markedPosition = 0
    self.elements = tuple()

  def getElements(self) : 
    return self.elements

  def getPosition(self) : 
    return self.position

  def hasNext(self) :
    result = False
    if self.position >= 0 and self.position < len(self.elements) :
      result = True
    else :
      result = False
    return result

  def isAfterLast(self) :
    result = False
    if self.position > len(self.elements) : 
      return True
    return result

  def isBeforeFirst(self) :
    result = False
    if self.position <= 0 : 
      return True
    return result

  def hasPrevious(self) :
    result = False
    if self.position > 1 and self.position <= len(self.elements) + 1 :
      result = True
    else :
      result = False
    return result

  def nextIndex(self) :
    result = 0
    result = self.position + 1
    return result

  def previousIndex(self) :
    result = 0
    result = self.position - 1
    return result

  def moveForward(self) :
    self.position = self.position + 1

  def moveBackward(self) :
    self.position = self.position - 1

  def moveTo(self, i) :
    self.position = i

  def setPosition(self, i) :
    self.position = i

  def markPosition(self) :
    self.markedPosition = self.position

  def movePosition(self, i) :
    self.position = i + self.position

  def moveToFirst(self) :
    self.position = 1

  def moveToLast(self) :
    self.position = len(self.elements)

  def moveToStart(self) :
    self.position = 0

  def moveToEnd(self) :
    self.position = len(self.elements) + 1

  def moveToMarkedPosition(self) :
    self.position = self.markedPosition

  def close(self) : 
    self.position = 0
    self.markedPosition = 0
    self.elements = tuple()

  def newOclIterator_Sequence(sq) :
    ot = OclIterator()
    ot.elements = tuple(sq)
    ot.position = 0
    return ot

  def getCurrent(self) :
    result = None
    if self.position <= len(self.elements) and self.position >= 1 : 
      result = (self.elements)[self.position - 1]
    return result

  def next(self) :
    result = None
    self.moveForward()
    return self.getCurrent()

  def previous(self) :
    result = None
    self.moveBackward()
    return self.getCurrent()

  def length(self) : 
    return len(self.elements)

  def at(self,i) : 
    return self.elements[i-1]


 




